File: generateParentheses/generateParentheses.js
/**
 * Given n pairs of parentheses, write a function to generate all combos of well-formed parantheses
 *
 * IOCE:
 * Input: integer n
 * Output: array of strings
 * Constraints: 1 <= n <= 8
 * Edge Cases: n/a
 *
 * Strategy
 * Stack
 *
 * Iterate from i = 0 to 2
 * add
 *
 * @param {number} n
 * @return {string[]}
 */

const generateParentheses = (n) => {

  let stack = [];
  let res = [];

  const innerFunction = (open = 0, closed = 0) => {

    if (closed === open && open === n) {
      res.push(stack.join(''));
      return;
    }

    if (open < n) {
      stack.push('(');
      innerFunction(open + 1, closed);
      stack.pop();
    }

    if (closed < open) {
      stack.push(')');
      innerFunction(open, closed + 1);
      stack.pop();
    }

  };

  innerFunction();

  return res;

};

let n = 3;
let expected = ['((()))', '(()())', '(())()', '()(())', '()()()'];
let result = generateParentheses(n);
console.log(generateParentheses(n), 'should equal', expected, ':', JSON.stringify(expected) === JSON.stringify(result));

n = 1;
expected = ['()'];
result = generateParentheses(n);
console.log(generateParentheses(n), 'should equal', expected, ':', JSON.stringify(expected) === JSON.stringify(result));

File: generateParentheses/generateParentheses.py
'''
Given n pairs of parantheses, write a function that generates all combos of well-formed parantheses.

IOCE:
Input: n, integer
Output: array of strings
Constraints: must be well-formed (i.e. valid)
Edge Cases: n/a

Strategy:
track count of closed and open parantheses
Closing parenthesis can only be added when # of open is greater than number of closed
if number of closed = number of open AND n, then the string is completed (base case)
Recursive approach

Pseudocode:
set variable closed to zero
set variable open to zero

if closed == open == n, then return array

if closed < open == n, then append closing parenthesis

if closed < open < n, then append a closing parenthesis and separately append an opening one, call recursive function

create inner recursive function which takes argument open, closed, and array of paranthesis

when return append result to result array
'''


def generate_parantheses(n: int):

  res = []
  stack = []

  def inner_function(open=0, closed=0):

    if open == closed == n:
      res.append(''.join(stack))
      return

    if open < n:
      stack.append('(')
      inner_function(open + 1, closed)
      stack.pop()

    if closed < open:
      stack.append(')')
      inner_function(open, closed + 1)
      stack.pop()

  inner_function()

  return res

n = 3
expected = ['((()))', '(()())', '(())()', '()(())', '()()()']
print(generate_parantheses(n))

n = 1
expected = ['()']
print(generate_parantheses(n))



